webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/reports/ComplianceLevel.scala

```diff
@@ -69,7 +74,114 @@ final case class CompliancePercent(
   , nonCompliant      : Double = 0
   , auditError        : Double = 0
   , badPolicyMode     : Double = 0
-)
+) {
+  val compliance = success+repaired+notApplicable+compliant+auditNotApplicable
+}
+
+object CompliancePercent {
+  /*
+   * Ensure that the sum is 100% and that no case disapear because it is
+   * less that 0.01%.
+   *
+   * Rules are:
+   * - always keeps at least 0.01% for any case
+   *   (given that we have 14 categories, it means that at worst, if 13 are at 0.1 in place of ~0%;
+   *   the last one will be false by 0.13%, which is ok)
+   * - the biggest value is rounded to compensate
+   * - in case of equality (ex: 1/3 success, 1/3 NA, 1/3 error), the biggest is
+   *   chosen accordingly to "getWorsteType" logic.
+   */
+  def fromLevels(c: ComplianceLevel): CompliancePercent = {
+    val MINPC = BigDecimal(0.01) // 0.01% since we keep two digits in percents
+
+    if(c.total == 0) {  // special case: let it be 0
+      CompliancePercent()
+    } else {
+      val total = c.total // not zero
+      def pc_for(i:Int) : Double = {
+        if(i == 0) {
+          0
+        } else {
+          val pc = (i * 100 / BigDecimal(total)).setScale(2, BigDecimal.RoundingMode.HALF_UP)
+          (if(pc < MINPC) MINPC else pc).toDouble
+        }
+      }
+
+      // default percent calculation
+      val levels = CompliancePercent.sortLevels(c)
+      // round the first ones keeping at least MINPC
+      val pc = levels.init.map { case (l, index) => (pc_for(l), index) }
+      // the last one is rounded by "100 - sum of other"
+      val pc_last = 100 - pc.map(_._1).sum
+
+      // check that the rounding error is below what is expected (0.15%) else warn
+      val error = Math.abs((pc_for(levels.last._1)-pc_last)/pc_last)
+      if( error > (levels.size+1)*MINPC) {
+        ComplianceLogger.info(s"Rounding error in compliances is above expected threshold: error is ${error} for ${c}")
+      }
+
+      // sort back percents by index
+      val all = ((pc_last, levels.last._2) :: pc).sortBy(_._2).map(_._1)
+
+      // create array of pecents
+      CompliancePercent.fromSeq(all)
+    }
+  }
+
+  /*
+   * returned a list of (level, index) where index correspond to the index of the
+   * level in the compliance array. The list is sorted in the relevant order so that
+   * the smaller level are at head, and in case of equality, the "worse order" order is
+   * kept (worse order last)
+   */
+  def sortLevels(c: ComplianceLevel): List[(Int, Int)] = {
+    // we want to compare accordingly to `ReportType.getWorsteType` but I don't see any
+    // way to do it directly since we don't use the same order in compliance.
+    // So we map index of a compliance element to it's worse type order and compare by index
+    val WORSE_ORDER = Seq(13, 4, 3, 12, 11, 2, 5, 6, 8, 0, 1, 9, 10, 7).zipWithIndex.toMap
```

Comment:
I can put that in report level and use it for the "get worse" algo, it would simplify both classes.
Here we would have name in the seq (Success.index, ehctc)

And we don't need a map, we have the index of the element, we can just access by index

Reply:
I did that, it feels OK to maintain